runtime/memprof.c

```diff
@@ -16,23 +16,531 @@
+#define Alloc_major(config)   Field(config, CONFIG_FIELD_ALLOC_MAJOR)
+#define Promote(config)       Field(config, CONFIG_FIELD_PROMOTE)
+#define Dealloc_minor(config) Field(config, CONFIG_FIELD_DEALLOC_MINOR)
+#define Dealloc_major(config) Field(config, CONFIG_FIELD_DEALLOC_MAJOR)
+
+/* Callback indexes. "Major" and "minor" are not distinguished here. */
+
+#define CB_NONE          0
+#define CB_ALLOC         1
+#define CB_PROMOTE       2
+#define CB_DEALLOC       3
+
+/* Maximum value of a callback index */
+#define CB_MAX           CB_DEALLOC
+
+/* How many bits required for a callback index */
+#define CB_BITS          2
+
+/* the mask for a given callback index */
+#define CB_MASK(cb) (1 << ((cb) - 1))
+
+/* Structure for each tracked allocation. Six words (with many spare
+ * bits in the final word). */
+
+struct entry_s {
+  /* Memory block being sampled. This is a weak GC root. Note that
+   * during the allocation callback of a block allocated directly by OCaml,
+   * this may be a comballoc offset. */
+  value block;
+
+  /* The value returned by the previous callback for this block, or
+   * the callstack (as a value-tagged pointer to the C heap) if the
+   * alloc callback has not been called yet.  This is a strong GC
+   * root. */
+  value user_data;
+
+  /* Number of samples in this block. */
+  size_t samples;
+
+  /* The size of this block. */
+  size_t wosize;
+
+  /* The thread currently running a callback for this entry,
+   * or NULL if there is none */
+  memprof_thread_t running;
+
+  /* Is `block` actually an offset? */
+  unsigned int offset : 1;
+
+  /* Was this block initially allocated in the minor heap? */
+  unsigned int alloc_young : 1;
+
+  /* The source of the allocation: normal allocations or custom_mem (SRC_*). */
+  unsigned int source : 2;
+
+  /* Has this block been promoted? Implies [alloc_young]. */
+  unsigned int promoted : 1;
+
+  /* Has this block been deallocated? */
+  unsigned int deallocated : 1;
+
+  /* Which callback (CB_*) is currently running for this entry. */
+  unsigned int callback : CB_BITS;
+
+  /* A mask of callbacks (1 << (CB_* - 1)) which have been called (not
+   * necessarily completed) for this entry. */
+  unsigned int callbacks : CB_MAX;
+
+  /* Has this entry been marked for deletion. */
+  unsigned int deleted : 1;
+
+  /* There are a number of spare bits here for future expansion,
+   * without increasing the size of an entry */
+};
+
+/* A resizable array of entry_s entries. */
+
+struct entries_s {
+  entry_t t; /* Pointer to array of entry_s structures */
+  size_t min_size, size, live; /* array allocation management */
+
+  /* Before this position, the [block] and [user_data] fields both
+   * point to the major heap ([young <= live]). */
+  size_t young;
+
+  /* There are no blocks to be evicted before this position
+   * ([evict <= live]). */
+  size_t evict;
+
+  /* There are no pending callbacks before this position
+   * ([next <= live]). */
+  size_t next;
```

Comment:
Hmm. I see that `next` isn't good, as that's usually a linked list field, but `pending_callback` is a bit verbose here. How about just `pending`, or something different from either, such as `look` (the next entry to look at)?

Reply:
The reason why I don't like `next` is rather that we don't know what "next thing" it is about. It could be the next thing to delete or the next entry with a young root.

`pending` is perhaps the best of your proposals because we already use the word "pending" everywhere for pending callbacks.

Reply:
Yeah, good.